repo/content/index/index.go

```diff
@@ -26,33 +25,40 @@ type Index interface {
 }
 
 // Open reads an Index from a given reader. The caller must call Close() when the index is no longer used.
-func Open(readerAt io.ReaderAt, v1PerContentOverhead uint32) (Index, error) {
-	h, err := v1ReadHeader(readerAt)
+func Open(data []byte, closer func() error, v1PerContentOverhead uint32) (Index, error) {
+	h, err := v1ReadHeader(data)
 	if err != nil {
 		return nil, errors.Wrap(err, "invalid header")
 	}
 
 	switch h.version {
 	case Version1:
-		return openV1PackIndex(h, readerAt, v1PerContentOverhead)
+		return openV1PackIndex(h, data, closer, v1PerContentOverhead)
 
 	case Version2:
-		return openV2PackIndex(readerAt)
+		return openV2PackIndex(data, closer)
 
 	default:
 		return nil, errors.Errorf("invalid header format: %v", h.version)
 	}
 }
 
-func readAtAll(ra io.ReaderAt, p []byte, offset int64) error {
-	n, err := ra.ReadAt(p, offset)
-	if n != len(p) {
-		return errors.Errorf("incomplete read at offset %v, got %v bytes, expected %v", offset, n, len(p))
-	}
+func safeSlice(data []byte, offset int64, length int) (v []byte, err error) {
```

Comment:
what are the semantics of the `data` slice returned by the mmap library? Do either `cap(data)` or `len(data)` reflect the addressable region?

Are there errors other than out-of-bounds errors that need to be considered?
If that's the case, then the error used for recovery may be misleading.

Reply:
the slice is sized to the length of memory-mapped segment, so the bounds checker than runs when reslicing will do the right thing.

Reply:
In that case, would checking the bounds work instead of trying to `panic` and `recover` ?